File: model_cpp/scalar_model.cpp
#include "scalar_model.h"
#include <stdlib.h>
#include <math.h>
#include <iostream>
#include <fstream>
#include <string>
#include <random>

using namespace std;
default_random_engine generator2(5);
uniform_real_distribution<double> uni_distribution2(0.0, 1.0);

ScalarModel::ScalarModel(char reward): reward(reward), cancer_cells(nullptr), healthy_cells(nullptr), time(0), glucose(0.0), oxygen(0.0), end_type('0'), init_hcell_count(0){
}

ScalarModel::~ScalarModel(){
    delete cancer_cells;
    delete healthy_cells;
}

void ScalarModel::reset(){
    delete cancer_cells;
    delete healthy_cells;
    HealthyCell::count = 0;
    CancerCell::count = 0;
    time = 0;
    glucose = 250000.0;
    oxygen = 2500000.0;
    healthy_cells = new CellList();
    cancer_cells = new CellList();
    for(int i = 0; i < 1000; i++)
        healthy_cells -> add(new HealthyCell('1'), 'h');
    cancer_cells -> add(new CancerCell('1'), 'c');
    go(350);
    init_hcell_count = HealthyCell::count;
}

void ScalarModel::cycle_cells(){
    int hcell_count = HealthyCell::count;
    int ccell_count = CancerCell::count;
    int count = hcell_count + ccell_count;
    CellNode * current_h = healthy_cells -> head;
    CellNode * current_c = cancer_cells -> head;
    CellNode * current;
    while(hcell_count > 0 || ccell_count > 0){
        if (rand() % (hcell_count + ccell_count) < ccell_count){
            ccell_count--;
            current = current_c;
            current_c = current_c -> next;
        } else {
            hcell_count--;
            current = current_h;
            current_h = current_h -> next;
        }
        cell_cycle_res result = current->cell->cycle(glucose, oxygen, count / 278);
        glucose -= result.glucose;
        oxygen -= result.oxygen;
        if (result.new_cell == 'h') //New healthy cell
            healthy_cells -> add(new HealthyCell('1'), 'h');
        else if (result.new_cell == 'c') // New cancer cell
            cancer_cells -> add(new CancerCell('1'), 'c');
    }
    healthy_cells -> deleteDeadAndSort();
    cancer_cells -> deleteDeadAndSort();
}

void ScalarModel::fill_sources(){
    glucose += 13000.0;
    oxygen += 450000.0;
}

void ScalarModel::go(int i){
    for(int x = 0; x < i; x++){
        time++;
        fill_sources();
        cycle_cells();
    }
}

void ScalarModel::irradiate(int dose){
    CellNode * current_h = healthy_cells -> head;
    while(current_h){
        current_h -> cell -> radiate(dose);
        current_h = current_h -> next;
    }
    healthy_cells -> deleteDeadAndSort();
    CellNode * current_c = cancer_cells -> head;
    while(current_c){
        current_c -> cell -> radiate(dose);
        current_c = current_c -> next;
    }
    cancer_cells -> deleteDeadAndSort();
}

double ScalarModel::act(int action){
    int dose = action + 1;
    int pre_hcell = HealthyCell::count;
    int pre_ccell = CancerCell::count;
    irradiate(dose);
    int m_hcell = HealthyCell::count;
    int m_ccell = CancerCell::count;
    go(24);
    int post_hcell = HealthyCell::count;
    int post_ccell = CancerCell::count;
    return adjust_reward(dose, pre_ccell - post_ccell, pre_hcell-min(post_hcell, m_hcell));
}

double ScalarModel::adjust_reward(int dose, int ccell_killed, int hcells_lost){
    //cout << dose << " " << reward << " " << ccell_killed << " " << hcells_lost << endl;
    if (inTerminalState() && reward != 'n'){
        if (end_type == 'L' || end_type == 'T'){
            return -1.0;
        } else{
            if (reward == 'd')
                return - (double) dose / 200.0 + 0.5 + (double) (HealthyCell::count) / 4000.0;
            else
                return 0.5 + (double) (HealthyCell::count) / 4000.0;
        }
    } else {
        if (reward == 'd' || reward == 'n')
            return - (double) dose / 200.0 + (double) (ccell_killed - 5.0* hcells_lost)/100000.0;
        else{
	    //cout << ccell_killed << " " << hcells_lost <<endl;
            return (double) (ccell_killed - (double) (reward - '0') * hcells_lost)/100000.0;
	    }
    }
}

bool ScalarModel::inTerminalState(){
    if (CancerCell::count <= 0){
        end_type = 'W';
        return true;
    } else if (HealthyCell::count < 10){
        end_type = 'L';
        return true;
    } else if (time > 1550){
        end_type = 'T';
        return true;
    } else {
        return false;
    }
}

TabularAgent::TabularAgent(ScalarModel * env, int cancer_cell_stages, int healthy_cell_stages, int actions, char state_type): env(env), cancer_cell_stages(cancer_cell_stages), healthy_cell_stages(healthy_cell_stages), actions(actions), state_type(state_type){
    Q_values = new double*[cancer_cell_stages * healthy_cell_stages];
    for(int i = 0; i < cancer_cell_stages * healthy_cell_stages; i++){
        Q_values[i] = new double[actions]();
    }
    if(state_type == 'o') { //log
        state_helper_hcells = exp(log(3500.0) / ((double) healthy_cell_stages - 2.0));
        state_helper_ccells = exp(log(40000.0) / ((double) cancer_cell_stages - 2.0));
    } else { // lin
        state_helper_hcells = 3500.0 / ((double) healthy_cell_stages - 2.0);
        state_helper_ccells = 40000.0 / ((double) cancer_cell_stages - 2.0);
    }
}

TabularAgent::~TabularAgent(){
    for(int i = 0; i < cancer_cell_stages * healthy_cell_stages; i++){
        delete[] Q_values[i];
    }
    delete[] Q_values;
}

int TabularAgent::state(){
    int hcell_state, ccell_state;
    if(state_type == 'o') { //log
        ccell_state = min(cancer_cell_stages - 1, (int) ceil(log(CancerCell::count + 1) / log(state_helper_ccells)));
        hcell_state = min(healthy_cell_stages - 1, (int) ceil(log(max(HealthyCell::count - 8, 1)) / log(state_helper_hcells)));
    } else{
        ccell_state = min(cancer_cell_stages - 1, (int) ceil((double) CancerCell::count / (double) state_helper_ccells) );
        hcell_state = min(healthy_cell_stages - 1, (int) ceil((double)  max(HealthyCell::count - 9, 0) / (double) state_helper_hcells) );
    }
    return ccell_state * healthy_cell_stages + hcell_state;
}

int TabularAgent::choose_action(int state, double epsilon){
    if(uni_distribution2(generator2) < epsilon) {
        return rand() % actions;
    } else {
        int max_ind = -1;
        double max_val = - 999999.0;
        for(int i = 0; i < actions; i++){
            if(max_val < Q_values[state][i]) {
                max_val = Q_values[state][i];
                max_ind = i;
            }
        }
        return max_ind;
    }
}

void TabularAgent::train(int steps, double alpha, double epsilon, double disc_factor){
    env -> reset();
    while(steps > 0){
        while (!env->inTerminalState() && steps > 0){
            int obs = state();
            int action = choose_action(obs, epsilon);
            double r = env->act(action);
            int new_obs = state();
            double max_val = - 99999.0;
            for(int i = 0; i < actions; i++){
                if(Q_values[new_obs][i] > max_val)
                    max_val = Q_values[new_obs][i];
            }
            Q_values[obs][action] = (1.0 - alpha) * Q_values[obs][action] + alpha * (r + disc_factor * max_val);
            steps--;
        }
        if(steps > 0)
            env -> reset();
    }
}

void TabularAgent::test(int episodes, bool verbose, double disc_factor, bool eval){
    double sum_scores = 0.0;
    double sum_error = 0.0;
    int sum_fracs = 0;
    int sum_doses = 0;
    int sum_w = 0;
    double survival = 0.0;
    for (int i = 0; i < episodes; i++){
        env -> reset();
        double sum_r = 0;
        double err = 0.0;
        int count = 0;
        int init_hcell = HealthyCell::count;
        while (!env->inTerminalState()){
            int obs = state();
            int action = choose_action(obs, 0.0);
            double r = env -> act(action);
            if (verbose)
                cout << action + 1 << " grays, reward =  " << r << endl;
            sum_fracs++;
            sum_doses += action + 1;
            sum_r += r;
            int new_obs = state();
            double max_val = - 99999.0;
            for(int i = 0; i < actions; i++){
                if(Q_values[new_obs][i] > max_val)
                    max_val = Q_values[new_obs][i];
            }
            err += pow(r + disc_factor * max_val - Q_values[obs][action], 2.0);
            count++;
        }
        if(verbose)
            cout << env -> end_type << endl;
        survival += (double) HealthyCell::count / (double) init_hcell;
        if (env -> end_type == 'W')
            sum_w++;
        sum_scores += sum_r;
        sum_error += err / (double) count;
    }
    cout << "Average score: " << sum_scores / (double) episodes << " MSE: " << sum_error / (double) episodes << endl;
    if(eval){
        cout << "TCP: " << 100.0 * (double) sum_w / (double) episodes << endl;
        cout << "Average num of fractions: " << (double) sum_fracs / (double) episodes << " Average dose: "  << (double) sum_doses / (double) episodes << " Average survival: " << survival / (double) episodes << endl;
    }
}

void TabularAgent::run(int n_epochs, int train_steps, int test_steps, double init_alpha, double alpha_mult, double init_epsilon, double end_epsilon, double disc_factor){
    test(test_steps, false, disc_factor, false);
    double alpha = init_alpha;
    double epsilon = init_epsilon;
    double epsilon_change = (double) (init_epsilon - end_epsilon) / (double) (n_epochs - 1);
    double alpha_change = (double) (init_alpha - alpha_mult) / (double) (n_epochs - 1);
    for(int i = 0; i < n_epochs; i++){
        cout << "Epoch " << i + 1 << endl;
        train(train_steps, alpha, epsilon, disc_factor);
        test(test_steps, false, disc_factor, false);
        alpha -= alpha_change;
        epsilon -= epsilon_change;
    }
}

void TabularAgent::treatment_var(int count){
    int** treatments = new int*[count];
    for(int i = 0; i < count; i++){
        treatments[i] = new int[100]();
    }
    for(int i = 0; i < count; i++){
        env -> reset();
        int j = 0;
        while (!env->inTerminalState()){
            int obs = state();
            int action = choose_action(obs, 0.0);
            double r = env -> act(action);
            treatments[i][j++] = action + 1;
        }
    }
    cout << "mean, std_error" << endl;
    for(int j = 0; j < 100; j++){
        int count_mean = 0;;
        int sum_mean = 0;
        for(int i = 0; i < count; i++){
            if (treatments[i][j] > 0){
                sum_mean += treatments[i][j];
                count_mean++;
            }
        }
        if (count_mean == 0)
            break;
        double mean = (double) sum_mean / (double) count_mean;
        double sum_std = 0.0;
        for(int i = 0; i < count; i++){
            if (treatments[i][j] > 0){
                sum_std += pow(((double) treatments[i][j]) - mean, 2.0);
            }
        }
        double std_error = sqrt(sum_std / (double) count_mean);
        cout << mean << ", " << std_error << endl;
    }
    for(int i = 0; i < count; i++){
        delete[] treatments[i];
    }
    delete[] treatments;
}

void TabularAgent::save_Q(string name){
    ofstream myfile;
    myfile.open(name);
    myfile << cancer_cell_stages << " " << healthy_cell_stages << " " << actions << "\n";
    for(int i = 0; i < cancer_cell_stages * healthy_cell_stages; i++){
        for(int j = 0; j < actions; j++){
            myfile << Q_values[i][j] << ", ";
        }
        myfile << "\n";
    }
    myfile.close();
}

void TabularAgent::load_Q(string name){
    ifstream f;
    f.open(name);
    if(!f.is_open()) throw std::runtime_error("Could not open file");
    string line;
    getline(f, line);
    int pos_space_1 = line.find(" ");
    int pos_space_2 = line.find(" ", pos_space_1);
    int t_cstage = stoi(line.substr(0, pos_space_1));
    int t_hstage = stoi(line.substr(pos_space_1 + 1, pos_space_2 - pos_space_1 - 1));
    int t_actions = stoi(line.substr(pos_space_2 + 1));
    if(t_cstage != cancer_cell_stages || t_hstage != healthy_cell_stages || t_actions != actions)
        throw std::runtime_error("Parameters do not match");
    getline(f, line);
    for(int i = 0; i < cancer_cell_stages * healthy_cell_stages; i++){
        for(int j = 0; j < actions; j++){
            getline(f, line, ',');
            Q_values[i][j] = stof(line);
        }
        getline(f, line);
    }
    f.close();
}

void no_treatment(){
    cout << "No treatment" << endl;
    ScalarModel * model = new ScalarModel('a');
    model -> reset();
    for(int i = 350; i < 2000; i += 50){
        cout << "Time: " << i << " Healthy cells: " <<  HealthyCell::count << " Cancer cells: " << CancerCell::count << endl;
        model -> go(50);
    }
    delete model;
}

void low_treatment(char reward){
    cout << "Low treatment" << endl;
    ScalarModel * model = new ScalarModel(reward);
    double sum_scores = 0.0;
    for(int i = 0; i < 25; i++){
        model -> reset();
        double sum_r = 0;
        int count = 0;
        while (!model->inTerminalState()){
            //int obs = state();
            sum_r += model -> act(0);
            count++;
        }
        sum_scores += sum_r;
    }
    cout << "Average reward " << (sum_scores / 25.0) << endl;
    delete model;
}

void baseline_treatment(char reward){
    cout << "Baseline treatment" << endl;
    ScalarModel * model = new ScalarModel(reward);
    double sum_scores = 0.0;
    for(int i = 0; i < 25; i++){
        model -> reset();
        double sum_r = 0;
        int count = 0;
        while (!model->inTerminalState()){
            //int obs = state();
            sum_r += model -> act(1);
            count++;
        }
        sum_scores += sum_r;
    }
    cout << "Average reward " << (sum_scores / 25.0) << endl;
    delete model;
}

void eval_baseline(char reward, int count){
    cout << "Baseline treatment" << endl;
    ScalarModel * model = new ScalarModel(reward);
    int sum_fracs = 0;
    int sum_doses = 0;
    int sum_w = 0;
    double survival = 0.0;
    for (int i = 0; i < count; i++){
        model -> reset();
        int count_f = 0;
        int init_hcell = HealthyCell::count;
        while (!env->inTerminalState()){
            int action = (count_f++ < 35)?1:-1;
            env -> act(action);
            sum_fracs++;
            sum_doses += action + 1;
        }
        survival += (double) HealthyCell::count / (double) init_hcell;
        if (env -> end_type == 'W')
            sum_w++;
    }
    cout << "TCP: " << 100.0 * (double) sum_w / (double) count << endl;
    cout << "Average num of fractions: " << (double) sum_fracs / (double) count << " Average dose: "  << (double) sum_doses / (double) count << " Average survival: " << survival / (double) count << endl;
    delete model;
}

void high_treatment(char reward){
    cout << "High treatment" << endl;
    ScalarModel * model = new ScalarModel(reward);
    double sum_scores = 0.0;
    for(int i = 0; i < 25; i++){
        model -> reset();
        double sum_r = 0;
        int count = 0;
        while (!model->inTerminalState()){
            //int obs = state();
            sum_r += model -> act(4);
            count++;
        }
        sum_scores += sum_r;
    }
    cout << "Average reward " << (sum_scores / 25.0) << endl;
    delete model;
}

void high_low_treatment(char reward){
    cout << "High low treatment" << endl;
    ScalarModel * model = new ScalarModel(reward);
    double sum_scores = 0.0;
    for(int i = 0; i < 25; i++){
        model -> reset();
        double sum_r = 0;
        int count = 0;
        while (!model->inTerminalState()){
            //int obs = state();
            if(count <= 3)
                sum_r += model -> act(3);
            else
                sum_r += model -> act(1);
            count++;
        }
        sum_scores += sum_r;
    }
    cout << "Average reward " << (sum_scores / 25.0) << endl;
    delete model;
}

void test_suite(char reward){
    low_treatment(reward);
    baseline_treatment(reward);
    high_treatment(reward);
    high_low_treatment(reward);
}


int main(int argc, char * argv[]){
    
    //cout << "Dose "<<endl;
    //test_suite('d');
        
    int n_epochs = stoi(argv[1]);
    char reward = argv[2][0];
    char state_type = argv[3][0];
    int cancer_cell_stages = stoi(argv[4]);
    int healthy_cell_stages = stoi(argv[5]);
    ScalarModel * model = new ScalarModel(reward);
    TabularAgent * agent = new TabularAgent(model, cancer_cell_stages, healthy_cell_stages, 5, state_type);
    if(argc == 8 && argv[7][0] == 'l')
        agent -> load_Q(argv[6]);
    //agent -> run(n_epochs, 5000, 10, 0.8, 0.05, 0.8, 0.01, 0.99);
    //agent -> test(5, true, 0.99, false);
    agent -> test(100, false, 0.99, true);
    agent -> treatment_var(100);
    //agent -> save_Q(argv[6]);
    delete model;
    delete agent;
}

File: model/scalar_model.py
from math import exp, log, ceil, floor
import matplotlib.pyplot as plt
from model.cell import CancerCell, HealthyCell, OARCell
import random
import numpy as np

alpha_tumor = 0.3
beta_tumor = 0.03
alpha_norm_tissue = 0.15
beta_norm_tissue = 0.03


class ScalarModel:

    def __init__(self, reward, draw=False):
        self.reward = reward
        self.draw_graph = draw
        self.cells = []

    def reset(self):
        del self.cells
        HealthyCell.cell_count = 0
        CancerCell.cell_count = 0
        self.time = 0
        self.cells = [CancerCell(0)] + [HealthyCell(0) for _ in range(1000)]
        self.glucose = 250000
        self.oxygen = 2500000
        if self.draw_graph:
            self.ccell_counts = [CancerCell.cell_count]
            self.hcell_counts = [HealthyCell.cell_count]
            self.ticks = [self.time]
        self.go(350)
        self.init_hcell_count = HealthyCell.cell_count

    def cycle_cells(self):
        to_add = []
        count = HealthyCell.cell_count + CancerCell.cell_count
        for cell in random.sample(self.cells, count):
            res = cell.cycle(self.glucose, count / 278 , self.oxygen)
            if len(res) > 2:  # If there are more than two arguments, a new cell must be created
                if res[2] == 0:  # Mitosis of a healthy cell
                    to_add.append(HealthyCell(0))
                elif res[2] == 1:
                    to_add.append(CancerCell(0))
            self.glucose -= res[0]  # The local variables are updated according to the cell's consumption
            self.oxygen -= res[1]
        self.cells = [cell for cell in self.cells if cell.alive] + to_add

    def fill_sources(self):
        self.glucose += 13000
        self.oxygen += 450000

    def go(self, ticks=1):
        for _ in range(ticks):
            self.time += 1
            self.fill_sources()
            self.cycle_cells()
            #graph
            if self.draw_graph:
                self.ticks.append(self.time)
                self.ccell_counts.append(CancerCell.cell_count)
                self.hcell_counts.append(HealthyCell.cell_count)

    def irradiate(self, dose):
        for cell in self.cells:
            cell.radiate(dose)
        self.cells = [cell for cell in self.cells if cell.alive]

    def observe(self):
        return HealthyCell.cell_count, CancerCell.cell_count

    def act(self, action):
        dose = action + 1
        pre_hcell = HealthyCell.cell_count
        pre_ccell = CancerCell.cell_count
        self.irradiate(dose)
        m_hcell = HealthyCell.cell_count
        m_ccell = CancerCell.cell_count
        self.go(24)
        post_hcell = HealthyCell.cell_count
        post_ccell = CancerCell.cell_count
        return self.adjust_reward(dose, pre_ccell - post_ccell, pre_hcell-min(post_hcell, m_hcell))

    def adjust_reward(self, dose, ccell_killed, hcells_lost):
        if self.inTerminalState():
            if self.end_type == "L" or self.end_type == "T":
                return -1
            else:
                if self.reward == 'dose':
                    return - dose / 200.0 + 0.5 + HealthyCell.cell_count / 4000.0
                else:
                    return 0.5 + HealthyCell.cell_count / 4000.0
        else:
            if self.reward == 'dose' or self.reward == 'oar':
                return - dose / 200.0 + (ccell_killed - 5.0 * hcells_lost) / 100000.0
            elif self.reward == 'killed':
                return  (ccell_killed - int(self.reward) * hcells_lost)/100000.0

    def inTerminalState(self):
        if CancerCell.cell_count <= 0 :
            self.end_type = 'W'
            return True
        elif HealthyCell.cell_count < 10:
            self.end_type = "L"
            return True
        elif self.time > 1550:
            self.end_type = "T"
            return True
        else:
            return False

    def draw(self, title):
        plt.plot(self.ticks, self.ccell_counts, 'r', label='Cancer cells')
        plt.plot(self.ticks, self.hcell_counts, 'b', label='Healthy cells')
        #plt.yscale('log')
        plt.xlabel('Hours')
        plt.ylabel("Count")
        plt.legend()
        plt.title(title)
        plt.savefig(''.join(title.lower().split()))


class TabularLearner:

    def __init__(self, env, cancer_cell_stages, healthy_cell_stages, actions, state_type):
        self.env = env
        self.cancer_cell_stages = cancer_cell_stages
        self.healthy_cell_stages = healthy_cell_stages
        self.actions = actions
        self.state_type = state_type
        self.Q = np.zeros((cancer_cell_stages, healthy_cell_stages, actions), dtype=float)
        if (state_type == 'o'): #log
            self.state_helper_hcells = exp(log(3500.0) / (healthy_cell_stages - 2.0))
            self.state_helper_ccells = exp(log(40000.0) / (cancer_cell_stages - 2.0))
        else: # lin
            self.state_helper_hcells = 3500.0 / (healthy_cell_stages - 2.0)
            self.state_helper_ccells = 40000.0 / (cancer_cell_stages - 2.0)

    def ccell_state(self, count):
        if (self.state_type == 'o'): # log
            return min(self.cancer_cell_stages - 1, int(ceil(log(CancerCell.cell_count + 1) / log(self.state_helper_ccells))))
        else:
            return min(self.cancer_cell_stages - 1, int(ceil(CancerCell.cell_count / self.state_helper_ccells)))


    def hcell_state(self, count):
        if (self.state_type == 'o'): # log
            return min(self.healthy_cell_stages - 1, ceil(log(max(HealthyCell.cell_count -8, 1) ) / log(self.state_helper_hcells)))
        else:
            return min(self.healthy_cell_stages - 1, int(ceil(max(HealthyCell.cell_count - 9, 0) / self.state_helper_hcells)))

    def convert(self, obs):
        return self.ccell_state(obs[1]), self.hcell_state(obs[0])

    def choose_action(self, state, epsilon):
        if random.random() < epsilon:
            return random.randint(0, self.actions - 1)
        else:
            return np.argmax(self.Q[state])

    def train(self, steps, alpha, epsilon, disc_factor):
        self.env.reset()
        while steps > 0:
            while not self.env.inTerminalState() and steps > 0:
                state = self.convert(self.env.observe())
                action = self.choose_action(state, epsilon)
                r = self.env.act(action)
                new_state = self.convert(self.env.observe())
                self.Q[state][action] = (1 - alpha) * self.Q[state][action] + alpha * (r + disc_factor * np.max(self.Q[new_state]))
                steps -= 1
            if steps > 0:
                self.env.reset()


    def test(self, episodes, disc_factor, verbose=False, eval=False):
        sum_scores = 0.0
        sum_error = 0.0
        lengths_arr = []
        fracs_arr = []
        doses_arr = []
        sum_w = 0
        survivals_arr = []
        for _ in range(episodes):
            self.env.reset()
            sum_r = 0
            err = 0.0
            count = 0
            fracs = 0
            doses = 0
            time = 0
            init_hcell = HealthyCell.cell_count
            while not self.env.inTerminalState():
                state = self.convert(self.env.observe())
                action = np.argmax(self.Q[state])
                r = self.env.act(action)
                if verbose:
                    print(action + 1, "grays, reward =", r)
                fracs += 1
                doses += action + 1
                time += 24
                sum_r += r
                new_state = self.convert(self.env.observe())
                err += pow(r + disc_factor * self.Q[new_state][action] - self.Q[state][action], 2.0)
                count += 1
            if verbose:
                print(self.env.end_type)
            if self.env.end_type == 'W':
                sum_w += 1
            if eval:
                fracs_arr.append(fracs)
                doses_arr.append(doses)
                lengths_arr.append(time)
                survival = HealthyCell.cell_count / init_hcell
                survivals_arr.append(survival)
            sum_scores += sum_r
            sum_error += err / count
        fracs_arr = np.array(fracs_arr)
        doses_arr = np.array(doses_arr)
        lengths_arr = np.array(lengths_arr)
        survivals_arr = np.array(survivals_arr)
        print("Average score:", sum_scores / episodes, "MSE:", sum_error / episodes)
        if eval:
            print("TCP: " , 100.0 *sum_w /episodes)
            print("Average num of fractions: ", np.mean(fracs_arr), " std dev: ", np.std(fracs_arr))
            print("Average radiation dose: ", np.mean(doses_arr), " std dev: ", np.std(doses_arr))
            print("Average duration: ", np.mean(lengths_arr), " std dev: ", np.std(lengths_arr))
            print("Average survival: ", np.mean(survivals_arr), " std dev: ", np.std(survivals_arr))


    def run(self, n_epochs, train_steps, test_steps, init_alpha, alpha_mult, init_epsilon, final_epsilon, disc_factor):
        self.test(test_steps, disc_factor)
        alpha = init_alpha
        epsilon = init_epsilon
        epsilon_change = (init_epsilon - final_epsilon) / (n_epochs - 1)
        alpha_change = (init_alpha - alpha_mult) / (n_epochs - 1)
        for i in range(n_epochs):
            print("Epoch ", i + 1)
            self.train(train_steps, alpha, epsilon, disc_factor)
            self.test(test_steps, disc_factor)
            alpha -= alpha_change
            epsilon -= epsilon_change

    def save_Q(self, name):
        np.save(name, self.Q, allow_pickle=False)

    def load_Q(self, name):
        self.Q = np.load(name+'.npy', allow_pickle=False)


random.seed(1234)
env = ScalarModel('killed', draw=True)
env.reset()
env.irradiate(15)
print(HealthyCell.cell_count, CancerCell.cell_count)
env.go()
env.draw('Single dose')
'''
agent = TabularLearner(env, 50, 5, 4)
agent.run(20, 2500, 100, 0.8, 0.5, 0.8, 0.05, 0.95)
agent.save_Q("Qvalkilled")
agent.test(100, verbose=True)
env = ScalarModel('dose')
agent = TabularLearner(env, 50, 5, 4)
agent.run(20, 2500, 100, 0.8, 0.5, 0.8, 0.05, 0.95)
agent.save_Q("Qvaldose")
agent.test(100, verbose=True)
'''




